docs/development/style_guide.md

```diff
@@ -0,0 +1,583 @@
+- **Composable:** Functions can be imported for use in other scripts
+
+❌ **Bad:** Code runs on import
+
+```python
+import sys
+import argparse
+
+# This runs immediately when imported, making testing difficult
+parser = argparse.ArgumentParser()
+parser.add_argument("--run-id", type=int, required=True)
+args = parser.parse_args()
+
+# Global side effects on import
+print(f"Fetching artifacts for run {args.run_id}")
+result = fetch_artifacts(args.run_id)
+print(f"Downloaded {len(result)} artifacts")
+```
+
+✅ **Good:** Separate definition from execution
+
+```python
+import sys
+import argparse
+from pathlib import Path
+
+
+def fetch_artifacts(run_id: int, output_dir: Path) -> list[Path]:
+    """Fetch artifacts from the given run ID."""
+    # Implementation here
+    return artifacts
+
+
+def main(argv: list[str]) -> int:
+    """Main entry point for the script."""
+    parser = argparse.ArgumentParser(description="Fetch artifacts from GitHub Actions")
+    parser.add_argument("--run-id", type=int, required=True)
+    parser.add_argument("--output-dir", type=Path, default=Path("artifacts"))
+
+    args = parser.parse_args(argv)
+
+    artifacts = fetch_artifacts(args.run_id, args.output_dir)
+    print(f"Downloaded {len(artifacts)} artifacts")
+    return 0
+
+
+if __name__ == "__main__":
+    sys.exit(main(sys.argv[1:]))
+```
+
+### Use named arguments for complicated function signatures
+
+Using positional arguments for functions that accept many arguments is error
+prone. Use keyword arguments to make function calls explicit and
+self-documenting.
+
+Benefits:
+
+- **Readability:** Clear what each argument represents at the call site
+- **Safety:** Prevents accidentally swapping arguments of the same type
+- **Maintainability:** Function signature can evolve without breaking calls
+
+> [!TIP]
+> Consider using named arguments when:
+>
+> - Function has more than 2-3 parameters
+> - Multiple parameters have the same type (especially booleans)
+> - The meaning of arguments isn't obvious from context
+
+❌ **Bad:** Positional arguments are error prone
+
+```python
+# What do these values mean? Easy to mix up the order
+result = build_artifacts(
+    "gfx942",
+    True,
+    False,
+    "/tmp/build",
+    ["rocblas", "hipblas"],
+)
+
+# Even worse: easy to swap boolean flags
+process_files(input_dir, output_dir, True, False, True)
+```
+
+✅ **Good:** Named arguments are clear and safe
+
+```python
+# Intent is immediately clear
+result = build_artifacts(
+    amdgpu_family="gfx942",
+    enable_testing=True,
+    use_ccache=False,
+    build_dir="/tmp/build",
+    components=["rocblas", "hipblas"],
+)
+
+# Flags are self-documenting
+process_files(
+    input_dir=input_dir,
+    output_dir=output_dir,
+    overwrite=True,
+    validate=False,
+    compress=True,
+)
+```
+
+## Bash guidelines
+
+> [!WARNING]
+> Bash is **strongly discouraged** for nontrivial usage in .yml GitHub Actions
```

Comment:
maybe add a comment that it also creates problem for the different platform support (win vs linux)